src/shared/augtool.cc

```diff
@@ -23,118 +23,164 @@
-            out += i.substr(pos+2);
+// execute cmd, returns raw output
+std::string augtool::get_cmd_out_raw(const std::string& cmd)
+{
+    std::lock_guard<std::mutex> lock(m_cmd_mutex);
+
+    std::string ret; // empty, default
+
+    if (m_process) {
+        auto command{cmd};
+        if (command.empty() || (command.back() != '\n')) {
+            command += "\n";
+        }
+
+        auto writeSuccess = m_process->write(command);
+        ret = m_process->readAllStandardOutput(500);
+        if (!writeSuccess) {
+            ret.clear();
         }
-        not_first = true;
     }
-    return out;
+    return ret;
 }
 
+// execute cmd, returns post-processed output
+std::string augtool::get_cmd_out(
+    const std::string& cmd,
+    bool key_value, // process only '<key> = <value>' output
+    const std::string& sep, // list separator
+    std::function<bool(std::string)> filter // returns true to exclude value
+)
+{
+    logDebug("augtool: '{}', key_value: {}, sep: '{}'", cmd, key_value, sep);
 
-std::string augtool::get_cmd_out_raw(std::string command) {
-    std::string ret;
-    bool err = false;
+    // execute the command
+    std::string cmdOutput = get_cmd_out_raw(cmd);
 
-    std::lock_guard<std::mutex> lock(mux);
-    if (init()) {
-        if(command.empty() || command.back() != '\n') {
-            command += "\n";
-        }
+    // split output into lines
+    std::vector<std::string> lines = fty::split(cmdOutput, "\n");
 
-        if(!prc->write(command)) {
-            err = true;
-        }
+    std::string ret; // empty, default
+
+    // expect at least 3 lines
+    if (lines.size() >= 3) {
+        // remove first and last prompts lines ("match...", "augtool>")
+        lines.erase(lines.begin());
+        lines.pop_back();
+
+        // built ret
+        for (auto line : lines) {
+            auto pos = line.find_first_of("=");
+            if (pos != std::string::npos) {
+                // extract value (right of '=' leaving 1st space)
+                line = line.substr(pos + 2);
+            }
+            else {
+                if (key_value) { // key_value only?
+                    continue; // ignore not '<key> = <value>' line
+                }
+            }
 
-        ret = prc->readAllStandardOutput(500);
-        return err ? "" : ret;
+            // apply exclusion filter
+            if (filter(line)) {
+                continue; // ignore line
+            }
+
+            if (line.empty()) { // inconsistent?
+                logDebug("adding an empty line!");
+            }
+
+            ret += (ret.empty() ? "" : sep) + line;
+        }
     }
-    return "";
-}
 
-void augtool::run_cmd(std::string cmd) {
-    get_cmd_out_raw(cmd);
+    logDebug("augtool: '{}', ret: '{}'", cmd, ret);
+    return ret;
 }
 
-static std::mutex clear_mux;
+// returns process instance
+augtool* augtool::get_instance(bool sudoer)
+{
+    static augtool instance_sudoer(true); // privileges escalation
```

Comment:
May be instanciate only one variable and work with a pointer

Reply:
static augtool instance_sudoer(true);
Make than only one instance is going to be instantiated for the full program